support/cas-server-support-duo/src/main/java/org/apereo/cas/adaptors/duo/config/DuoSecurityAuthenticationEventExecutionPlanConfiguration.java

```diff
@@ -112,25 +112,29 @@ public VariegatedMultifactorAuthenticationProvider duoMultifactorAuthenticationP
 
     @Bean
     public Action prepareDuoWebLoginFormAction() {
-        return new PrepareDuoWebLoginFormAction(duoMultifactorAuthenticationProvider(), applicationContext);
+        return new PrepareDuoWebLoginFormAction();
     }
 
     @Bean
     public Action determineDuoUserAccountAction() {
-        return new DetermineDuoUserAccountAction(duoMultifactorAuthenticationProvider(), applicationContext);
+        return new DetermineDuoUserAccountAction();
+    }
+
+    @Bean
+    public Action duoInitializeLoginAction() {
+        return new DuoInitializeLoginAction(duoMultifactorAuthenticationProvider());
     }
 
     @Bean
     @RefreshScope
     public AuthenticationMetaDataPopulator duoAuthenticationMetaDataPopulator() {
-        final String authenticationContextAttribute = casProperties.getAuthn().getMfa().getAuthenticationContextAttribute();
-        return new AuthenticationContextAttributeMetaDataPopulator(authenticationContextAttribute, duoAuthenticationHandler(),
-            duoMultifactorAuthenticationProvider());
+        return new DuoAuthenticationMetaDataPopulator(casProperties.getAuthn().getMfa().getAuthenticationContextAttribute(),
+                duoAuthenticationHandler());
     }
 
     @RefreshScope
     @Bean
-    public AuthenticationHandler duoAuthenticationHandler() {
+    public DuoAuthenticationHandler duoAuthenticationHandler() {
```

Comment:
I understand why you had to do this, but this needs to be re-worked. We should try to stick to the most generic type possible, and introduce things that can be polymorphically recognized.